src/node_realm-inl.h

```diff
@@ -115,18 +119,31 @@ int64_t Realm::base_object_count() const {
   return base_object_count_;
 }
 
+// V8 can not infer reference cycles between global persistent handles, e.g.
+// the Realm's Context handle and the per-realm function handles.
+// Attach the per-realm strong persistent values' lifetime to the context's
+// global object so that they are kept alive when the context is still
+// reachable.
 #define V(PropertyName, TypeName)                                              \
   inline v8::Local<TypeName> Realm::PropertyName() const {                     \
-    return PersistentToLocal::Strong(PropertyName##_);                         \
+    return PersistentToLocal::StrongUnChecked(PropertyName##_);                \
   }                                                                            \
   inline void Realm::set_##PropertyName(v8::Local<TypeName> value) {           \
     PropertyName##_.Reset(isolate(), value);                                   \
+    if (!value.IsEmpty()) {                                                    \
+      PropertyName##_.SetWeak();                                               \
+      v8::Local<v8::Context> ctx = context();                                  \
+      ctx->Global()                                                            \
```

Comment:
I think this should be moved to a separate method that only gets run for shadow realms. The overhead here is unnecessary for the main context.

Reply:
Done.